vue files - include parametric embedding support for javascript and styles

haidubogdan opened the pull request:
This pull request will include updates for embedding support for vue files.

- javscript

![image](https://github.com/user-attachments/assets/58c84b44-599b-498e-88b5-584e2085ac6b)

- pug | jade

![image](https://github.com/user-attachments/assets/6f4634ff-c17b-463a-9853-049fc2206975)


- scss style language

![image](https://github.com/user-attachments/assets/1c44a6d9-d975-4bd9-8ed7-0f2b5a4e680a)


- less style language

![image](https://github.com/user-attachments/assets/c985cdc1-d315-40c3-9e85-06ead462f036)

For the moment, I didn't manage to include embedding for **typescript**.
The solution might not be that straightforward, as it doesn't have a standard lexer.

---
**^Add meaningful description above**

<details open>
<summary>Click to collapse/expand PR instructions</summary>

By opening a pull request you confirm that, unless explicitly stated otherwise, the changes -

 - are all your own work, and you have the right to contribute them.
 - are contributed solely under the terms and conditions of the Apache License 2.0 (see section 5 of the license for more information).

Please make sure (eg. `git log`) that all commits have a valid name and email address for you in the Author field.

If you're a first time contributor, see the Contributing guidelines for more information.

If you're a committer, please label the PR before pressing "Create pull request" so that the right test jobs can run.

### PR approval and merge checklist:

1. [x] Was this PR [correctly labeled](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=240884239#PRsandYouAreviewerGuide-PRtriggeredCIJobs(conditionalCIpipeline)), did the right tests run? When did they run?
2. [x] Is this PR [squashed](https://cwiki.apache.org/confluence/display/NETBEANS/git%3A+squash+and+merge)?
3. [x] Are author name / email address correct? Are [co-authors](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors#creating-co-authored-commits-on-the-command-line) correctly listed? Do the commit messages need updates?
3. [x] Does the PR title and description still fit after the Nth iteration? Is the description sufficient to appear in the release notes?

If this PR targets the delivery branch: [don't merge](https://cwiki.apache.org/confluence/display/NETBEANS/Pull+requests+for+delivery). ([full wiki article](https://cwiki.apache.org/confluence/display/NETBEANS/PRs+and+You+-+A+reviewer+Guide))

</details>

matthiasblaesing commented on ide/css.lib/src/org/netbeans/modules/css/lib/nbparser/CssParser.java:
```diff
@@ -93,12 +93,23 @@ public void parse(Snapshot snapshot, Task task, SourceModificationEvent event) t
             //parse just an empty string in case of an oversize snapshot
             CharSequence source = tooLargeSnapshot ? "" : snapshot.getText();
-            
+
+            String mimePath = snapshot.getMimePath() != null ? snapshot.getMimePath().getPath() : null;
+            if (mimePath != null) {
+                //vue plugin embedding
+                if (mimePath.startsWith("text/x-vue/text/scss")) { //NOI18N
+                    mimeType = "text/scss"; //NOI18N
+                } else if (mimePath.startsWith("text/x-vue/text/less")){ //NOI18N
+                    mimeType = "text/less"; //NOI18N
+                }
+            }
+
```
I think we should fix the CSS parsing correctly. Currently CSS/SCSS/LESS switching is a mess as it relies on three different patterns. Before this change:

- for tests the mimetype is forced via the constructor `org.netbeans.modules.css.lib.nbparser.CssParser.CssParser(String)`
- for "plain" stylesheets it is deduced from the file mimetype
- for all other types parsing falls hard to `text/css`

What I envision (please see if want to pursue this, else I'll create a followup) is, that we drop the specialcases like this:

- check the inner most element of the mimepath (`snapshot.getMimePath().getMimeType(snapshot.getMimePath().size() - 1)`)
- if that is from the set `text/less`, `text/scss` or `text/css` use it, else fall back to `text/css`

For unittests we should be able to create a correct snapshot, for normal execution, the behaviour should match.

That way we don't need to special case CSS/SCSS/LESS embeddings in other languages and the new vue-specialcase is not necessary.

haidubogdan replied:
I discovered that I had to use `snapshot.getMimePath().getMimeType(snapshot.getMimePath().size() - 2)` for the penultimate mimeType.

matthiasblaesing commented on webcommon/javascript2.vue/src/org/netbeans/modules/javascript2/vue/editor/embedding/VueHtmlEmbeddingProvider.java:
```diff
@@ -61,12 +62,16 @@ public List<Embedding> getEmbeddings(final Snapshot snapshot) {
                 if (id.equals(VueTokenId.HTML)) {
                     embeddings.add(snapshot.create(ts.offset(), t.length(), TARGET_MIME_TYPE));
+                } else if (t.text() != null) {
+                    //issue with bracematcher
+                    String fake = new String(new char[t.text().toString().length()]).replace("\0", FILLER); //NOI18N
+                    embeddings.add(snapshot.create(fake, TARGET_MIME_TYPE));
```
It would be great to have an example what goes wrong here. I have an idea, but like to verify. Thank you.

haidubogdan replied:
Yes, here is an example.
I see that it's mostly for javascript **script** tag.
Maybe a mimepath bug also?

![image](https://github.com/user-attachments/assets/f651339d-0e0b-480b-af8c-142140311302)

matthiasblaesing replied:
Yeah, most probably a matcher bug.... This makes this change specially interesting as it makes it necessary to cleanup the language embeddings.

haidubogdan replied:
The script part is easy to fix but for :

![image](https://github.com/user-attachments/assets/d1499347-21db-4f5c-8883-dbc5defb180a)

Seems to be a issue with the fact that there is no offset check.

Unfortunately the issue is already in the vue implementation from nb 26-rc1, but somehow I didn't notice it.

matthiasblaesing commented on webcommon/javascript2.vue/src/org/netbeans/modules/javascript2/vue/editor/lexer/VueLexer.java:
```diff
@@ -29,6 +29,13 @@
+    public static final String PUG_LANG = "pug"; //NOI18N
+    public static final String JADE_LANG = "jade"; //NOI18N
+    
+    //Styles
+    //embedding not working in netbeans ide
+    public static final String SCSS_LANG = "scss"; //NOI18N
+    public static final String LESS_LANG = "less"; //NOI18N
```
With the suggested changes for `CssParser` it should be possible to create a mimepath `text/x-vue/text/scss`, that will be properly handled. This might reveals broken assumptions in the CSS parser, but then we should fix them ;-)

haidubogdan replied:
Maybe I left the comment in a work in progress moment.
The token separation is useful for the embedding process.

matthiasblaesing commented on webcommon/javascript2.vue/src/org/netbeans/modules/javascript2/vue/resources/vue16.png:
Nicer. I asumme you are still the author?

haidubogdan replied:
Yes, it's my creation :) .

matthiasblaesing commented on webcommon/javascript2.vue/test/unit/data/testfiles/embedding/vue_embedding_01.vue:
```diff
@@ -0,0 +1,51 @@
+<!--
```
For tests please include the ASF header. There is no need to skip it and we should carry it where possible. This comment applies for other test files too.

matthiasblaesing reviewed (commented):
Thank you. Looks good to me. I left a few comments inline.

matthiasblaesing commented on ide/css.lib/src/org/netbeans/modules/css/lib/nbparser/CssParser.java:
```diff
@@ -81,24 +81,40 @@ public void parse(Snapshot snapshot, Task task, SourceModificationEvent event) t
-        String mimeType = topLevelSnapshotMimetype != null ? topLevelSnapshotMimetype : (fo == null ? null : fo.getMIMEType());
+        String fileName;
+        String mimeType;
+        if (fo != null) {
+            fileName= fo.getPath();
+            MimePath mimePath = snapshot.getMimePath();
+
+            //scss, less mime embedding is a top layer for text/css
+            //positioned at the penultimate position
+            if (mimePath != null && mimePath.size() > 2) {
+                mimeType = mimePath.getMimeType(mimePath.size() - 2);
+            } else {
+                mimeType = fo.getMIMEType();
+            }
+        } else {
+            fileName = "no file";
+            mimeType = topLevelSnapshotMimetype;
+        }
```
```suggestion
        if (topLevelSnapshotMimetype == null) {
            fileName= fo.getPath();
            MimePath mimePath = snapshot.getMimePath();

            //scss, less mime embedding is a top layer for text/css
            //positioned at the penultimate position
            if (mimePath != null && mimePath.size() > 2) {
                mimeType = mimePath.getMimeType(mimePath.size() - 2);
            } else {
                mimeType = snapshot.getMimeType();
            }
        } else {
            fileName = "no file";
            mimeType = topLevelSnapshotMimetype;
        }
```

haidubogdan replied:
@matthiasblaesing So far I see that the `topLevelSnapshotMimetype` is a flag used in  **Test Context** to parse css **Source** Snapshot without a fileObject and also imposing a mimeType. 

My observation with the current logic is that there is no real check if a **fo** is null or not, yet it's still possible to use a **Source** object without a **FileObject** and have a null **topLevelSnapshotMimetype** .
So it will have an exception at this part.

```
fileName= fo.getPath();
```

Strangely, the update didn't break any tests, so topLevelSnapshotMimetype must be used for sources without a fileObject. My general idea would be that the mimeType should be obtained correctly from the fileObject.

Maybe a issue could be with the testUtils which creates a common text/css mimeType document to be used for parsing 
 scss and less files, so the context needs to be created with specific flags.
The coupling scss | less => css mimeType is really confusing here :) .

matthiasblaesing commented on webcommon/javascript2.vue/src/org/netbeans/modules/javascript2/vue/editor/embedding/VueHtmlEmbeddingProvider.java:
```diff
@@ -56,17 +61,26 @@ public List<Embedding> getEmbeddings(final Snapshot snapshot) {
-                if (id.equals(VueTokenId.HTML)) {
-                    embeddings.add(snapshot.create(ts.offset(), t.length(), TARGET_MIME_TYPE));
+                if (token.text() != null && id instanceof VueTokenId) {
+                    switch ((VueTokenId) id) {
+                        //javascript, css, scss, less require content embedding due to braces matcher issue
+                        case HTML, JAVASCRIPT, CSS -> {
+                            embeddings.add(snapshot.create(ts.offset(), token.length(), TARGET_MIME_TYPE));
+                        }
+                        case STYLE_SCSS, STYLE_LESS -> {
+                            String fake = new String(new char[token.text().length()]).replace("\0", FILLER); //NOI18N
+                            embeddings.add(snapshot.create(fake, TARGET_MIME_TYPE));
+                        }
+                    }
```
This looks strange. The VueHtmlEmbeddingProvider should provide embedding for HTML. The other embeddings should be handled/and are being handled by their respecitve counter parts.

haidubogdan replied:
Added your update proposal.

matthiasblaesing commented on webcommon/javascript2.vue/src/org/netbeans/modules/javascript2/vue/editor/embedding/VueJsEmbeddingProvider.java:
```diff
@@ -0,0 +1,83 @@
+import org.netbeans.modules.parsing.api.Snapshot;
+import org.netbeans.modules.parsing.spi.EmbeddingProvider;
+
+/**
+ *
+ *
+ * @author bhaidu
+ */
```
Please remove. The author information is available from git history.

matthiasblaesing commented on webcommon/javascript2.vue/src/org/netbeans/modules/javascript2/vue/editor/lexer/VueTokenId.java:
```diff
@@ -37,8 +40,12 @@
 public enum VueTokenId implements TokenId {
     HTML("html"), // NOI18N
     CSS("css"), // NOI18N
+    STYLE_SCSS("css"), // NOI18N
+    STYLE_LESS("css"), // NOI18N
```
Will this scale? Webdevelopers have the tendency to reinvent the wheel (LESS, SCSS, SASS and now also JS variants).

haidubogdan replied:
It's the easiest approach to assign embedding on the tokens.

It's mostly a vue template syntax, so it's also hard to see a generic implementation on this part, and for the moment I do rely on what coloring lexer exists in netbeans platform.

haidubogdan replied:
@matthiasblaesing you are right, the **lang** logic could use some thought on scaling.

Recently I found about `stylus`  preprocessor syntax.

So maybe a refactor could have used a dynamic approach.
Set in the tokenId primaryCategory the expected mime  ex: (text/stylus)

In the embedding function

```
protected LanguageEmbedding<? extends TokenId> embedding(

String mime = tokenId.primaryCategory()
language = (Language<? extends TokenId>) provider.findLanguage(mime ); 
```

matthiasblaesing commented on webcommon/javascript2.vue/src/org/netbeans/modules/javascript2/vue/grammar/antlr4/coloring/ColoringLexerAdaptor.java:
```diff
@@ -76,4 +61,43 @@ public void setVarInterpolationOpened(boolean state) {
+    }
+
+    private String extractLangFromInput() {
+        String input = this.getText();
+        String langAttrEq = LANG_ATTR + "=";  //NOI18N
+        if (input == null || !input.startsWith(langAttrEq)) {
+            return null;
+        }
+        String langValue = input.substring(langAttrEq.length());
+        //check if only quotes
+        if (langValue.length() == 2) {
+            return null;
+        }
+
+        return langValue.substring(1, langValue.length() - 1);
```
Is this safe? HTML allows unquoted attributes and there maybe spaces between attribute, equals sign and value. Is the same true for vue?

haidubogdan replied:
As it is specific to vue syntax, I think it should be ok.

matthiasblaesing commented on webcommon/javascript2.vue/src/org/netbeans/modules/javascript2/vue/grammar/antlr4/coloring/VueAntlrColoringLexer.g4:
```diff
@@ -85,16 +100,34 @@ SCRIPT_ATTR_QUOTE : '"' {this.setAttrQuoteState(true);}->type(QUOTE_ATTR);
 
+mode INSIDE_STYLE_TAG_START;
+
+STYLE_LANG_ATTR : 'lang=' StringLiteral {this.setStyleLanguage();}->type(HTML);
```
see comment about HTML5 structure above.

mbien commented on webcommon/javascript2.vue/src/org/netbeans/modules/javascript2/vue/editor/VueLanguage.java:
```diff
@@ -93,6 +93,13 @@ public Parser getParser() {
 
+    @Override
+    public boolean isIdentifierChar(char c) {
+        return super.isIdentifierChar(c)
+                || c == '-' //NOI18N
+                ;
```
this could be one line I think ;)

haidubogdan commented:
@matthiasblaesing , @mbien  as a work in progress info I will remove some embeddings implementation in favor of a custom lexer plugin as we have with angular plugin.

matthiasblaesing commented:
@haidubogdan I had another quick look and I noticed, that JS integration did not work properly. I fixed it be cutting down the `VueHtmlEmbeddingProvider` to the scope it should handle (in my first pass I already left an inline comment, that it looked strange). With this code:

```java
    @Override
    public List<Embedding> getEmbeddings(final Snapshot snapshot) {
        TokenHierarchy<?> tokenHierarchy = snapshot.getTokenHierarchy();
        TokenSequence<?> ts = tokenHierarchy.tokenSequence();

        if (ts == null || !ts.isValid()) {
            return Collections.emptyList();
        }
        ts.moveStart();
        List<Embedding> embeddings = new ArrayList<>();

        try {
            while (ts.moveNext()) {
                Token<?> token = ts.token();
                TokenId id = token.id();

                if (token.text() != null && id == VueTokenId.HTML) {
                    embeddings.add(snapshot.create(ts.offset(), token.length(), TARGET_MIME_TYPE));
                }
            }
        } catch (Exception ex) {
            return Collections.emptyList();
        }

        if (embeddings.isEmpty()) {
            return Collections.emptyList();
        } else {
            return Collections.singletonList(Embedding.create(embeddings));
        }
    }
```

Opening `component_01.vue` seems to work correctly. The JS highlighting and code completion worked. The same is true for the style tag.

So why not wrap this up and make it a mergable change, so that we can get it in?

matthiasblaesing commented:
@haidubogdan thank you! Looks like a nice improvement. One final adjustment request. Would you please have a look here:

- https://github.com/matthiasblaesing/netbeans/tree/pr-8401
- https://github.com/apache/netbeans/commit/4bd90afc6b3b9cbc79f8ae6d40424970182acfda

There is no need to open the `css.prep` and `javascript2.jade` modules. You only want to access the language and for that you can use the lexer api to load them at runtime based on mimetype. The new unittest ensures, that the languages don't go away silently.

Would you check the change and see if you can integrate it?

haidubogdan commented:
@matthiasblaesing 
Beside the decoupling code,
I've removed the css prep and jade dependency from the project, but kept them on test level.
I've added a simple lexer test for pug script.
There was a typo variable on script language lexer state, which I fixed.

matthiasblaesing commented:
> @matthiasblaesing Beside the decoupling code, I've removed the css prep and jade dependency from the project, but kept them on test level. 

This must be reverted. The dependency is required, so that when `javascript2.vue` module is loaded, `css.prop` and `javascript2.jade` are made available. If not some other module loads these two modules (or they are auto activating), then `Language.find` will not be able to resolve the language and you will run into problems, because you pass a `null` language to `LanguageEmbedding.create`. Adding the too modules as test modules essentially invalidates the unittest, as now testing and runtime test different environments.

The are two levels of coupling here:
- General module dependency: required, as we assume the target languages are present. We are not interested in the internals, but need the modules to be available.
- Implementation/code dependency: We don't need access to the source code. It was only ever used to load the language and usage can be substituted.

>I've added a simple lexer test for pug script. There was a typo variable on script language lexer state, which I fixed.

Makes sense.

Looking at the github diff, the revert of the changes to `webcommon/javascript2.jade/nbproject/project.xml` is incomplete, there are still whitespace changes.

haidubogdan commented:
@matthiasblaesing 
I've re-added the dependency for css prep and jade modules.
